multicast/multicast.go

```diff
@@ -0,0 +1,354 @@
+	addrs := ciliumNode.Spec.Addresses
+	var ciliumInternalIP v2.NodeAddress
+	for _, addr := range addrs {
+		if addr.AddrType() == addressing.NodeCiliumInternalIP {
+			ip, err := netip.ParseAddr(addr.IP)
+			if err != nil {
+				continue
+			}
+			if ip.Is4() {
+				ciliumInternalIP = addr
+			}
+		}
+	}
+	if ciliumInternalIP.IP == "" {
+		return v2.NodeAddress{}, fmt.Errorf("ciliumInternalIP not found")
+	}
+	return ciliumInternalIP, nil
+}
+
+func (m *Multicast) listCiliumPods(ctx context.Context) ([]corev1.Pod, error) {
+	ciliumPods, err := m.client.ListPods(ctx, m.params.CiliumNamespace, metav1.ListOptions{LabelSelector: "k8s-app=cilium"})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list Cilium pods: %w", err)
+	}
+	return ciliumPods.Items, nil
+}
+
+// ListGroup lists multicast groups in every node
+func (m *Multicast) ListGroup() error {
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+
+	ciliumPods, err := m.listCiliumPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	for _, ciliumPod := range ciliumPods {
+		go func(pod corev1.Pod) {
+			defer wg.Done()
+			// List multicast groups
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "group", "list"}
+			output, err := m.client.ExecInPod(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			output_string := "Node: " + pod.Spec.NodeName + "\n" + output.String()
+			fmt.Fprintln(m.params.Writer, output_string)
+		}(ciliumPod)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	var err_ret error
+	for fetchdata := range errCh {
+		if fetchdata != nil {
+			err_ret = errors.Join(err_ret, fetchdata)
+		}
+	}
+	return err_ret
+}
+
+// ListSubscriber lists multicast subscribers in every node for the specified multicast group or all multicast groups
+func (m *Multicast) ListSubscriber() error {
+	if m.params.MulticastGroupIP == "" && !m.params.All {
+		return fmt.Errorf("group-ip or all flag must be specified")
+	} else if m.params.MulticastGroupIP != "" && m.params.All {
+		return fmt.Errorf("only one of group-ip or all flag must be specified")
+	}
+
+	var target string
+	if m.params.All {
+		target = "all"
+	} else {
+		target = m.params.MulticastGroupIP
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), m.params.WaitDuration)
+	defer cancel()
+
+	ciliumPods, err := m.listCiliumPods(ctx)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	for _, ciliumPod := range ciliumPods {
+		go func(pod corev1.Pod) {
+			defer wg.Done()
+			// List multicast subscribers
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "list", target}
+			output, err := m.client.ExecInPod(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+			//ToDo:detect error type and handle properly
```

Comment:
nit: If you are going to leave this `//ToDo: ` here, which is okay, it should be formatted as `// TODO `. Same for other occurrences of `ToDo`, also inline comments should start like `// <comment>` rather than `//<comment>`.